x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/related_integrations/related_integrations.test.tsx

```diff
@@ -0,0 +1,694 @@
+    });
+  });
+
+  describe('removing an item', () => {
+    it('removes just added item', async () => {
+      render(<TestForm />, { wrapper: createReactQueryWrapper() });
+
+      await addRelatedIntegrationRow();
+      await removeRelatedIntegrationRow();
+
+      expect(screen.queryByTestId(RELATED_INTEGRATION_ROW)).not.toBeInTheDocument();
+    });
+
+    it('removes just added item after integration was selected', async () => {
+      render(<TestForm />, { wrapper: createReactQueryWrapper() });
+
+      await addRelatedIntegrationRow();
+      await selectFirstEuiComboBoxOption({
+        comboBoxToggleButton: screen.getByTestId(COMBO_BOX_TOGGLE_BUTTON_TEST_ID),
+      });
+      await removeRelatedIntegrationRow();
+
+      expect(screen.queryByTestId(RELATED_INTEGRATION_ROW)).not.toBeInTheDocument();
+    });
+
+    it('submits undefined when all just added integrations removed', async () => {
+      const handleSubmit = jest.fn();
+
+      render(<TestForm onSubmit={handleSubmit} />, { wrapper: createReactQueryWrapper() });
+
+      await addRelatedIntegrationRow();
+      await selectFirstEuiComboBoxOption({
+        comboBoxToggleButton: screen.getByTestId(COMBO_BOX_TOGGLE_BUTTON_TEST_ID),
+      });
+      await removeRelatedIntegrationRow();
+      await submitForm();
+
+      expect(handleSubmit).toHaveBeenCalledWith({
+        data: undefined,
+        isValid: true,
+      });
+    });
+
+    it('submits undefined after previously saved integrations were removed', async () => {
+      const initialRelatedIntegrations: RelatedIntegration[] = [
+        { package: 'package-a', version: '^1.2.3' },
+      ];
+      const handleSubmit = jest.fn();
+
+      render(<TestForm initialState={initialRelatedIntegrations} onSubmit={handleSubmit} />, {
+        wrapper: createReactQueryWrapper(),
+      });
+
+      await waitForIntegrationsToBeLoaded();
+      await removeRelatedIntegrationRow();
+      await submitForm();
+
+      expect(handleSubmit).toHaveBeenCalledWith({
+        data: undefined,
+        isValid: true,
+      });
+    });
+  });
+});
+
+interface TestFormProps {
+  initialState?: RelatedIntegration[];
+  onSubmit?: (args: { data: RelatedIntegration[]; isValid: boolean }) => void;
+}
+
+function TestForm({ initialState, onSubmit }: TestFormProps): JSX.Element {
+  const { form } = useForm({
+    options: { stripEmptyFields: false },
+    schema: {
+      relatedIntegrationsField: {
+        type: FIELD_TYPES.JSON,
+      },
+    },
+    defaultValue: {
+      relatedIntegrationsField: initialState,
+    },
+    onSubmit: async (formData, isValid) =>
+      onSubmit?.({ data: formData.relatedIntegrationsField, isValid }),
+  });
+
+  return (
+    <Form form={form} component="form">
+      <RelatedIntegrations path="relatedIntegrationsField" />
+      <button type="button" onClick={form.submit}>
+        {'Submit'}
+      </button>
+    </Form>
+  );
+}
+
+function waitForIntegrationsToBeLoaded(): Promise<void> {
```

Comment:
I like the declarative nature of these helpers; it looks/feels very much like a cypress test (although much more lightweight). It would be similarly nice to abstract/reuse these in other test files, but I'm struggling to find/think of a nice way to do this. We have mocks in `[moduleName].mock.ts`; maybe helpers would go in `[moduleName].test.helpers.ts` ?

Reply:
I also had it mind but also struggled with a common folder for them. I plan to discuss it with the team on our tech time meeting and come up with some ideas. To be honest I was inspired by EUI tests since similar helper utility functions used there. Ideally speaking we should have two things here

- Generic EUI utility test functions like `openComboBox()` exposed by EUI
- Our business domain utility functions like `waitForMyComponentToBeLoaded()`

`waitForIntegrationsToBeLoaded()` depends on the loaded state the 'RelatedIntegrationField' component sets for the combobox. It's not a generic thing but it could be reused for rule creation testing.

I'd like to address any changes in a separate PR after I collect more ideas and decision is clear.

Reply:
+1 to extracting away EUI utility test functions. I could reuse these in my Related Integrations tests.